server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java

```diff
@@ -505,6 +515,17 @@ private void indexValue(DocumentParserContext context, String value) {
             return;
         }
 
+        // Check that a dimension field is single-valued and not an array
+        if (dimension && context.doc().getByKey(fieldType().name()) != null) {
+            throw new IllegalArgumentException("Dimension field [" + fieldType().name() + "] cannot be a multi-valued field.");
+        }
+
+        if (dimension && value.length() > DIMENSION_MAX_LENGTH) {
+            throw new IllegalArgumentException(
+                "Dimension field [" + fieldType().name() + "] cannot be more than [" + DIMENSION_MAX_LENGTH + "] characters long."
+            );
+        }
```

Comment:
Could you do the assertion on the length of the `binaryValue` instead? That's the length we really care about.